README.md
# React Native WebImage

An image component for react-native with persistent disk and memory caching.

It is mostly a wrapper around native libraries which actually do the hard work.
On Android it uses [Glide](https://github.com/bumptech/glide), on iOS –
[SDWebImage](https://github.com/rs/SDWebImage).

:warning: The package is currently in alpha stage of development. If you find a bug or missing functionality, please feel free to report, but better fix or implement what you want and send a pull request to GitHub repository.

## Installation

```sh
npm install --save react-native-web-image
react-native link
```

| version       | react-native |
| ------------- | ------------ |
| 0.0.5 - 0.0.6 | &gt;=0.47.0  |
| 0.0.3 - 0.0.4 | &gt;=0.40.0  |
| &lt;=0.0.2    | &lt;0.40.0 (tested with 0.37.0) |

## Usage

```javascript
import React, { Component } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import WebImage from 'react-native-web-image'

export default class App extends Component {
  render() {
    const imageUri = 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'
    return (
      <View style={styles.container}>
        <Text style={styles.welcome}>Sample image</Text>
        <View style={styles.imgContainer}>
          <WebImage style={styles.img} source={{uri:imageUri}}/>
        </View>
      </View>
    )
  }
}

const white = '#FFFFFF'
const blue = 'rgb(0,0,255)'
const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: white,
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  img: {
    flex: 1,
  },
  imgContainer: {
    flexDirection: 'row',
    flex: 1,
    borderWidth: 1,
    borderColor: blue,
  }
})
```

## API

### WebImage element

```javascript
<WebImage source={source}/>
```

| Attribute  | Type     | Description |
| ---------- | -------- | ----------- |
| source     | Object   | Describes image source (mimics original Image element) |
| source.uri | String   | (**Required**) URL of the image |
| resizeMode | Enum{'cover','contain','stretch','center'} | Determine resize mode for image. Default: 'contain' |
| onError    | Function | Will be called on error |

#### Resize modes

* **contain** - Scale the image uniformly (maintain the image's aspect ratio)
  so that both dimensions (width and height) of the image will be equal to or
  less than the corresponding dimension of the view (minus padding).
* **cover** - Scale the image uniformly so that both dimensions (width and
  height) of the image will be equal to or larger than the corresponding
  dimension of the view (minus padding).
* **stretch** - Scale width and height independently, aspect ration will not be
  preserved.
* **center** - Scale the image down so that it is completely visible, if bigger
  than the area of the view. The image will not be scaled up. The image will be
  placed at the center of the view.

#### onError(event)

* `event.error` - String representation of error (platform dependent)
* `event.uri` - URI which leads to error

## Author

[Vladimir Timofeev](https://github.com/vovkasm)

## Contributors

* [iisue](https://github.com/iisue)

## License

* Main source code is licensed under the MIT License.
* SDWebImage (embedded in project) is licensed under the MIT License.
* Glide (included via gradle) is licensed under Apache 2.0 License.


CHANGELOG.md
# Changelog

## 0.0.6 (2017-10-04)

* [fix] (iOS) Add vendored SDWebImage to Header Search Paths, should fix archive (#8)

## 0.0.5 (2017-09-09)

* [feature] (Android) Implement borders & paddings (#2 - only Android part)
* [fix] Migrate to prop-types
* [fix] (Android) Getting a Random crash (#6)
* [fix] (Android) NullPointer Exception (#4)
* [chore] npm can't pull submodules, so embed SDWebImage back into project (#8)
* [chore] BREAKING: Update to React Native 0.48.2 (Sorry I don't know what minimum compatible version for now, it did work with 0.47 at least)
* [chore] Update SDWebImage to 4.1.0
* [chore] Update glide to 3.8.0
* [chore] Improve SampleApp

## 0.0.4 (2017-01-15)

* [feature] onError handler

## 0.0.3 (2017-01-13)

* [feature] Update to React Native 0.40.0
* [feature] Add sample application (for easy hacking)
* [fix] Reduce package size

## 0.0.2 (2017-01-06)

* [feature] Add resizeMode support (#1)

## 0.0.1 (2016-11-19)

* [feature] basic functionality


samples/SampleApp/samples.js
import PropTypes from 'prop-types'
import React from 'react'
import { ScrollView, StyleSheet, Text, View } from 'react-native'
import WebImage from 'react-native-web-image' // 'react-native-web-image' sources here

const Sample = (props) => {
  const { title, width, height, ...other } = props
  const containerStyle = {}
  if (width !== undefined) {
    containerStyle.width = width
  }
  if (height !== undefined) {
    containerStyle.height = height
  }
  return <View style={styles.container}>
    <Text style={styles.welcome}>{title}</Text>
    <View style={[styles.imgContainer, containerStyle]}>
      <WebImage style={styles.img} {...other} />
    </View>
  </View>
}
Sample.propTypes = {
  title: PropTypes.string.isRequired,
  width: PropTypes.number,
  height: PropTypes.number,
  resizeMode: PropTypes.string
}

const Samples = (props) => {
  return <ScrollView>
    <Sample
      title='[200x150] in [auto x 120] block'
      source={{uri: 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'}}
      height={120}
    />
    <Sample
      title='[200x150] in [100 x 100] block'
      source={{uri: 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'}}
      height={100} width={100}
    />
    <Sample
      title='[200x150] in [auto x 120] block, resizeMode=cover'
      source={{uri: 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'}}
      resizeMode='cover'
      height={120}
    />
    <Sample
      title='[200x150] in [100 x 100] block, resizeMode=cover'
      source={{uri: 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'}}
      resizeMode='cover'
      height={100} width={100}
    />
    <Sample
      title='[200x150] in [auto x 120] block, resizeMode=stretch'
      source={{uri: 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'}}
      resizeMode='stretch'
      height={120}
    />
    <Sample
      title='[200x150] in [100 x 100] block, resizeMode=stretch'
      source={{uri: 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'}}
      resizeMode='stretch'
      height={100} width={100}
    />
    <Sample
      title='[200x150] in [auto x 120] block, resizeMode=center'
      source={{uri: 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'}}
      resizeMode='center'
      height={120}
    />
    <Sample
      title='[200x150] in [100 x 100] block, resizeMode=center'
      source={{uri: 'https://placeholdit.imgix.net/~text?txtsize=33&txt=200x150&w=200&h=150'}}
      resizeMode='center'
      height={100} width={100}
    />
    <Sample
      title='image with onError handler, see logs'
      source={{uri: 'https://httpbin.org/status/404'}}
      width={100} height={100}
      onError={(e) => { console.log('WebImage onError handler: ', e) }}
    />
  </ScrollView>
}

export default Samples

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#FFFFFF'
  },
  welcome: {
    fontSize: 12,
    textAlign: 'center',
    margin: 10
  },
  img: {
    flex: 1,
    borderColor: 'rgb(127,255,127)'
  },
  imgContainer: {
    flexDirection: 'row',
    borderWidth: 1,
    borderColor: 'rgb(0,0,255)'
  }
})
